### PR TITLE
add smallvec for multiset join

### PR DESCRIPTION
**Pull request by zzlk**
joins are a nested structure, they are hash maps with vecs in them. We already re-added Clear impls everywhere to avoid freeing the hash table memory, but when clear() is called it still deletes all the values in the hash map, which causes the vecs to drop and free their memory. If we use a smallvec instead then, at least for the paxos case, most of the allocations never happen and also none of the subsequent frees happen either. This results in a pretty decent speedup for the paxos test.

### CODE REVIEW

**Comment by zzlk** [comment]
at least in our micro benchmarks, perf only improves from this change :)

<details>
```
arithmetic/timely       time:   [38.722 ms 39.200 ms 39.755 ms]
                        change: [-0.4098% +1.2023% +2.8224%] (p = 0.16 > 0.05)
                        No change in performance detected.
Found 15 outliers among 100 measurements (15.00%)
  6 (6.00%) high mild
  9 (9.00%) high severe

arithmetic/pipeline     time:   [837.58 ms 843.37 ms 850.15 ms]
                        change: [+0.0124% +0.7676% +1.6672%] (p = 0.06 > 0.05)
                        No change in performance detected.
Found 13 outliers among 100 measurements (13.00%)
  7 (7.00%) high mild
  6 (6.00%) high severe

arithmetic/iter         time:   [798.36 µs 803.80 µs 810.49 µs]
                        change: [-7.4694% -5.3173% -3.3454%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 10 outliers among 100 measurements (10.00%)
  4 (4.00%) high mild
  6 (6.00%) high severe

arithmetic/iter-collect time:   [6.4823 ms 6.5098 ms 6.5418 ms]
                        change: [-18.760% -16.860% -14.988%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 11 outliers among 100 measurements (11.00%)
  5 (5.00%) high mild
  6 (6.00%) high severe

arithmetic/raw          time:   [17.088 ms 17.205 ms 17.336 ms]
                        change: [-30.171% -26.913% -23.632%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 8 outliers among 100 measurements (8.00%)
  6 (6.00%) high mild
  2 (2.00%) high severe

arithmetic/hydroflow/compiled
                        time:   [288.50 µs 295.63 µs 304.67 µs]
                        change: [-0.0554% +1.4749% +3.3561%] (p = 0.08 > 0.05)
                        No change in performance detected.
Found 14 outliers among 100 measurements (14.00%)
  1 (1.00%) low severe
  2 (2.00%) low mild
  5 (5.00%) high mild
  6 (6.00%) high severe

arithmetic/hydroflow/compiled_no_cheating
                        time:   [6.3290 ms 6.8920 ms 7.6002 ms]
                        change: [-3.6002% +5.5393% +16.672%] (p = 0.33 > 0.05)
                        No change in performance detected.
Found 16 outliers among 100 measurements (16.00%)
  1 (1.00%) low mild
  2 (2.00%) high mild
  13 (13.00%) high severe

arithmetic/hydroflow/surface
                        time:   [5.9503 ms 6.0253 ms 6.1277 ms]
                        change: [-10.034% -4.4292% +0.0720%] (p = 0.11 > 0.05)
                        No change in performance detected.
Found 8 outliers among 100 measurements (8.00%)
  1 (1.00%) low mild
  3 (3.00%) high mild
  4 (4.00%) high severe

fan_in/hydroflow        time:   [142.76 ms 143.47 ms 144.30 ms]
                        change: [-7.9113% -6.6294% -5.4401%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 11 outliers among 100 measurements (11.00%)
  5 (5.00%) high mild
  6 (6.00%) high severe

fan_in/hydroflow/surface
                        time:   [126.82 ms 127.80 ms 129.08 ms]
                        change: [-1.5093% -0.3762% +0.7301%] (p = 0.54 > 0.05)
                        No change in performance detected.
Found 12 outliers among 100 measurements (12.00%)
  4 (4.00%) high mild
  8 (8.00%) high severe

fan_in/timely           time:   [68.277 ms 68.754 ms 69.330 ms]
                        change: [-2.9702% -2.0173% -1.0542%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 9 outliers among 100 measurements (9.00%)
  2 (2.00%) high mild
  7 (7.00%) high severe

fan_in/iters            time:   [5.6665 ms 5.6804 ms 5.6970 ms]
                        change: [-0.3947% -0.0705% +0.2737%] (p = 0.69 > 0.05)
                        No change in performance detected.
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe

fan_in/loops            time:   [8.4234 ms 8.4536 ms 8.4860 ms]
                        change: [-1.0820% -0.4640% +0.1550%] (p = 0.13 > 0.05)
                        No change in performance detected.
Found 8 outliers among 100 measurements (8.00%)
  7 (7.00%) high mild
  1 (1.00%) high severe

fan_out/hydroflow/scheduled
                        time:   [78.743 ms 79.080 ms 79.541 ms]
                        change: [-0.8266% -0.1106% +0.5788%] (p = 0.77 > 0.05)
                        No change in performance detected.
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe

fan_out/hydroflow/surface
                        time:   [5.4947 ms 5.5479 ms 5.6129 ms]
                        change: [-0.4238% +1.2677% +3.0641%] (p = 0.15 > 0.05)
                        No change in performance detected.
Found 6 outliers among 100 measurements (6.00%)
  6 (6.00%) high severe

fan_out/timely          time:   [50.488 ms 50.654 ms 50.836 ms]
                        change: [-3.2304% -2.2457% -1.2510%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) high mild
  1 (1.00%) high severe

fan_out/sol             time:   [5.3142 ms 5.4054 ms 5.5096 ms]
                        change: [-0.0375% +1.8199% +4.0224%] (p = 0.09 > 0.05)
                        No change in performance detected.
Found 10 outliers among 100 measurements (10.00%)
  2 (2.00%) high mild
  8 (8.00%) high severe

fork_join/hydroflow     time:   [7.9273 ms 7.9472 ms 7.9711 ms]
                        change: [-3.7618% -2.3173% -1.0766%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 7 outliers among 100 measurements (7.00%)
  3 (3.00%) high mild
  4 (4.00%) high severe

fork_join/hydroflow_builder
                        time:   [7.1470 ms 7.1678 ms 7.1938 ms]
                        change: [-0.4139% +0.1231% +0.6356%] (p = 0.66 > 0.05)
                        No change in performance detected.
Found 6 outliers among 100 measurements (6.00%)
  3 (3.00%) high mild
  3 (3.00%) high severe

fork_join/timely        time:   [6.7127 ms 6.7633 ms 6.8226 ms]
                        change: [-0.7926% +0.2132% +1.1750%] (p = 0.68 > 0.05)
                        No change in performance detected.
Found 8 outliers among 100 measurements (8.00%)
  1 (1.00%) high mild
  7 (7.00%) high severe

fork_join/raw           time:   [5.6737 ms 5.6968 ms 5.7228 ms]
                        change: [-0.7770% +0.1274% +0.8699%] (p = 0.78 > 0.05)
                        No change in performance detected.
Found 14 outliers among 100 measurements (14.00%)
  4 (4.00%) high mild
  10 (10.00%) high severe

identity/timely         time:   [45.156 ms 45.288 ms 45.455 ms]
                        change: [-0.6350% -0.1674% +0.3140%] (p = 0.50 > 0.05)
                        No change in performance detected.
Found 5 outliers among 100 measurements (5.00%)
  2 (2.00%) high mild
  3 (3.00%) high severe

identity/pipeline       time:   [839.21 ms 843.21 ms 848.55 ms]
                        change: [+0.1807% +0.7317% +1.4271%] (p = 0.02 < 0.05)
                        Change within noise threshold.
Found 12 outliers among 100 measurements (12.00%)
  10 (10.00%) high mild
  2 (2.00%) high severe

identity/iter           time:   [6.6704 ms 6.7321 ms 6.8090 ms]
                        change: [+0.3799% +1.3380% +2.4399%] (p = 0.01 < 0.05)
                        Change within noise threshold.
Found 9 outliers among 100 measurements (9.00%)
  3 (3.00%) high mild
  6 (6.00%) high severe

identity/iter-collect   time:   [17.862 ms 18.005 ms 18.166 ms]
                        change: [+1.3497% +2.3105% +3.2822%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 6 outliers among 100 measurements (6.00%)
  5 (5.00%) high mild
  1 (1.00%) high severe

identity/raw            time:   [18.352 ms 18.745 ms 19.149 ms]
                        change: [+5.3411% +7.3146% +9.8347%] (p = 0.00 < 0.05)
                        Performance has regressed.

identity/hydroflow      time:   [18.914 ms 19.059 ms 19.267 ms]
                        change: [-0.7651% +0.2957% +1.5414%] (p = 0.64 > 0.05)
                        No change in performance detected.
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe

identity/hydroflow/compiled
                        time:   [5.8651 ms 5.9445 ms 6.0674 ms]
                        change: [-0.3695% +1.2083% +3.2322%] (p = 0.23 > 0.05)
                        No change in performance detected.
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) high mild
  1 (1.00%) high severe

identity/hydroflow/surface
                        time:   [5.7661 ms 5.8330 ms 5.9124 ms]
                        change: [-13.698% -9.1711% -4.8129%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 5 outliers among 100 measurements (5.00%)
  1 (1.00%) high mild
  4 (4.00%) high severe

join/usize/usize/timely time:   [37.299 ms 38.284 ms 39.428 ms]
                        change: [-10.564% -5.0257% +0.4831%] (p = 0.09 > 0.05)
                        No change in performance detected.
Found 10 outliers among 100 measurements (10.00%)
  4 (4.00%) high mild
  6 (6.00%) high severe

join/usize/usize/sol    time:   [34.947 ms 35.130 ms 35.321 ms]
                        change: [-19.076% -16.359% -13.799%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild

join/String/String/timely
                        time:   [77.180 ms 77.726 ms 78.319 ms]
                        change: [-18.381% -16.351% -14.424%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 13 outliers among 100 measurements (13.00%)
  9 (9.00%) high mild
  4 (4.00%) high severe

join/String/String/sol  time:   [75.197 ms 75.949 ms 76.777 ms]
                        change: [+2.1438% +3.6663% +5.2372%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 5 outliers among 100 measurements (5.00%)
  4 (4.00%) high mild
  1 (1.00%) high severe

micro/ops/identity      time:   [7.8552 µs 7.8612 µs 7.8670 µs]
                        change: [-3.4530% -2.2684% -1.2412%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 5 outliers among 100 measurements (5.00%)
  1 (1.00%) high mild
  4 (4.00%) high severe

micro/ops/unique        time:   [47.854 µs 47.931 µs 48.008 µs]
                        change: [-18.928% -14.734% -10.372%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 5 outliers among 100 measurements (5.00%)
  1 (1.00%) high mild
  4 (4.00%) high severe

micro/ops/map           time:   [9.0153 µs 9.0457 µs 9.0795 µs]
                        change: [-7.6565% -5.6669% -3.7903%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 7 outliers among 100 measurements (7.00%)
  5 (5.00%) high mild
  2 (2.00%) high severe

micro/ops/flat_map      time:   [19.180 µs 19.248 µs 19.323 µs]
                        change: [-6.2490% -4.1129% -2.1453%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 10 outliers among 100 measurements (10.00%)
  6 (6.00%) high mild
  4 (4.00%) high severe

micro/ops/join          time:   [115.32 µs 116.67 µs 118.24 µs]
                        change: [-5.7639% -3.2167% -1.0942%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 15 outliers among 100 measurements (15.00%)
  2 (2.00%) high mild
  13 (13.00%) high severe

micro/ops/difference    time:   [100.94 µs 101.29 µs 101.68 µs]
                        change: [-7.9412% -3.3055% -0.0393%] (p = 0.12 > 0.05)
                        No change in performance detected.
Found 7 outliers among 100 measurements (7.00%)
  4 (4.00%) high mild
  3 (3.00%) high severe

micro/ops/merge         time:   [18.431 µs 18.769 µs 19.215 µs]
                        change: [-1.9437% +1.6710% +5.7295%] (p = 0.42 > 0.05)
                        No change in performance detected.
Found 8 outliers among 100 measurements (8.00%)
  4 (4.00%) high mild
  4 (4.00%) high severe

micro/ops/tee           time:   [8.0269 µs 8.0994 µs 8.1929 µs]
                        change: [-10.051% -6.9272% -4.1021%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 12 outliers among 100 measurements (12.00%)
  4 (4.00%) high mild
  8 (8.00%) high severe

micro/ops/fold          time:   [35.027 µs 35.219 µs 35.457 µs]
                        change: [-1.5699% -0.2556% +1.0225%] (p = 0.71 > 0.05)
                        No change in performance detected.

micro/ops/sort          time:   [135.46 µs 136.17 µs 137.03 µs]
                        change: [-2.4017% -1.1499% -0.0963%] (p = 0.05 < 0.05)
                        Change within noise threshold.
Found 5 outliers among 100 measurements (5.00%)
  5 (5.00%) high mild

micro/ops/crossjoin     time:   [157.51 µs 157.74 µs 157.99 µs]
                        change: [-1.5474% -0.9649% -0.4565%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 7 outliers among 100 measurements (7.00%)
  4 (4.00%) high mild
  3 (3.00%) high severe

micro/ops/anti_join     time:   [13.374 µs 13.419 µs 13.469 µs]
                        change: [-0.6803% -0.1526% +0.4444%] (p = 0.60 > 0.05)
                        No change in performance detected.
Found 4 outliers among 100 measurements (4.00%)
  4 (4.00%) high severe

micro/ops/next_tick/small
                        time:   [79.724 µs 80.127 µs 80.652 µs]
                        change: [-2.0738% -1.1602% -0.1972%] (p = 0.02 < 0.05)
                        Change within noise threshold.
Found 19 outliers among 100 measurements (19.00%)
  4 (4.00%) low severe
  5 (5.00%) low mild
  7 (7.00%) high mild
  3 (3.00%) high severe

micro/ops/next_tick/big time:   [87.262 µs 87.736 µs 88.308 µs]
                        change: [-2.5502% -1.0419% +0.5465%] (p = 0.20 > 0.05)
                        No change in performance detected.
Found 16 outliers among 100 measurements (16.00%)
  3 (3.00%) low severe
  7 (7.00%) low mild
  3 (3.00%) high mild
  3 (3.00%) high severe

micro/ops/group_by      time:   [12.038 µs 12.104 µs 12.164 µs]
                        change: [-1.4336% -0.5978% +0.1862%] (p = 0.15 > 0.05)
                        No change in performance detected.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild

reachability/timely     time:   [3.3187 ms 3.3336 ms 3.3499 ms]
                        change: [-0.0341% +0.7441% +1.4207%] (p = 0.05 > 0.05)
                        No change in performance detected.
Found 10 outliers among 100 measurements (10.00%)
  8 (8.00%) high mild
  2 (2.00%) high severe

reachability/differential
                        time:   [24.653 ms 24.868 ms 25.117 ms]
                        change: [-1.1118% +0.2035% +1.3798%] (p = 0.77 > 0.05)
                        No change in performance detected.
Found 8 outliers among 100 measurements (8.00%)
  2 (2.00%) low mild
  2 (2.00%) high mild
  4 (4.00%) high severe

reachability/hydroflow/scheduled
                        time:   [2.3595 ms 2.3825 ms 2.4109 ms]
                        change: [-0.4868% +0.6965% +1.9260%] (p = 0.28 > 0.05)
                        No change in performance detected.
Found 13 outliers among 100 measurements (13.00%)
  4 (4.00%) high mild
  9 (9.00%) high severe

reachability/hydroflow  time:   [2.3188 ms 2.3231 ms 2.3278 ms]
                        change: [-1.9952% -1.4449% -0.9694%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) high mild
  1 (1.00%) high severe

reachability/hydroflow/surface
                        time:   [6.3868 ms 6.4492 ms 6.5548 ms]
                        change: [-0.1776% +1.0742% +2.9595%] (p = 0.18 > 0.05)
                        No change in performance detected.
Found 7 outliers among 100 measurements (7.00%)
  4 (4.00%) high mild
  3 (3.00%) high severe

reachability/hydroflow/surface_cheating
                        time:   [1.7176 ms 1.7269 ms 1.7387 ms]
                        change: [-0.6978% +0.1203% +1.0432%] (p = 0.79 > 0.05)
                        No change in performance detected.
Found 5 outliers among 100 measurements (5.00%)
  1 (1.00%) high mild
  4 (4.00%) high severe

symmetric_hash_join/no_match
                        time:   [118.62 µs 118.78 µs 118.96 µs]
                        change: [-5.5418% -3.5023% -1.6562%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 8 outliers among 100 measurements (8.00%)
  1 (1.00%) low mild
  4 (4.00%) high mild
  3 (3.00%) high severe

symmetric_hash_join/match_keys_diff_values
                        time:   [166.90 µs 167.23 µs 167.63 µs]
                        change: [+0.0839% +0.5295% +0.9225%] (p = 0.01 < 0.05)
                        Change within noise threshold.
Found 13 outliers among 100 measurements (13.00%)
  7 (7.00%) high mild
  6 (6.00%) high severe

symmetric_hash_join/match_keys_same_values
                        time:   [166.84 µs 167.28 µs 167.82 µs]
                        change: [-0.5184% -0.1673% +0.1231%] (p = 0.33 > 0.05)
                        No change in performance detected.
Found 9 outliers among 100 measurements (9.00%)
  3 (3.00%) high mild
  6 (6.00%) high severe

symmetric_hash_join/zipf_keys_low_contention_unique_values
                        time:   [151.59 µs 152.52 µs 153.54 µs]
                        change: [-2.6947% -1.2209% +0.1396%] (p = 0.11 > 0.05)
                        No change in performance detected.
Found 5 outliers among 100 measurements (5.00%)
  1 (1.00%) low mild
  2 (2.00%) high mild
  2 (2.00%) high severe

symmetric_hash_join/zipf_keys_high_contention_unique_values
                        time:   [3.5810 ms 3.6149 ms 3.6576 ms]
                        change: [-1.7211% -0.2775% +1.1011%] (p = 0.71 > 0.05)
                        No change in performance detected.
Found 5 outliers among 100 measurements (5.00%)
  1 (1.00%) high mild
  4 (4.00%) high severe

upcase_in_place/timely  time:   [25.221 ms 25.282 ms 25.354 ms]
                        change: [-1.3376% -0.2848% +0.4789%] (p = 0.62 > 0.05)
                        No change in performance detected.
Found 13 outliers among 100 measurements (13.00%)
  5 (5.00%) high mild
  8 (8.00%) high severe

upcase_allocating/timely
                        time:   [76.234 ms 76.574 ms 77.065 ms]
                        change: [-0.7253% -0.0994% +0.5782%] (p = 0.78 > 0.05)
                        No change in performance detected.
Found 4 outliers among 100 measurements (4.00%)
  1 (1.00%) high mild
  3 (3.00%) high severe

concatting/timely       time:   [54.162 ms 54.464 ms 54.921 ms]
                        change: [-0.1973% +0.9579% +2.0606%] (p = 0.10 > 0.05)
                        No change in performance detected.
Found 9 outliers among 100 measurements (9.00%)
  7 (7.00%) high mild
  2 (2.00%) high severe

upcase_in_place/raw     time:   [21.654 ms 21.809 ms 21.971 ms]
                        change: [-1.0200% -0.1325% +0.8317%] (p = 0.78 > 0.05)
                        No change in performance detected.
Found 27 outliers among 100 measurements (27.00%)
  15 (15.00%) low severe
  1 (1.00%) low mild
  6 (6.00%) high mild
  5 (5.00%) high severe

upcase_allocating/raw   time:   [78.917 ms 79.212 ms 79.575 ms]
                        change: [+2.2509% +3.4443% +4.4068%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 8 outliers among 100 measurements (8.00%)
  5 (5.00%) high mild
  3 (3.00%) high severe

concatting/raw          time:   [49.471 ms 50.443 ms 51.702 ms]
                        change: [+1.7533% +3.7941% +6.5019%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 13 outliers among 100 measurements (13.00%)
  2 (2.00%) high mild
  11 (11.00%) high severe

upcase_in_place/iter    time:   [6.5196 ms 6.5302 ms 6.5424 ms]
                        change: [-1.3334% -0.5971% -0.0553%] (p = 0.06 > 0.05)
                        No change in performance detected.
Found 5 outliers among 100 measurements (5.00%)
  1 (1.00%) high mild
  4 (4.00%) high severe

upcase_allocating/iter  time:   [68.470 ms 68.775 ms 69.186 ms]
                        change: [-1.2707% -0.6689% -0.0376%] (p = 0.04 < 0.05)
                        Change within noise threshold.
Found 6 outliers among 100 measurements (6.00%)
  1 (1.00%) low mild
  2 (2.00%) high mild
  3 (3.00%) high severe

concatting/iter         time:   [24.134 ms 24.220 ms 24.331 ms]
                        change: [-2.4469% -1.3594% -0.3725%] (p = 0.01 < 0.05)
                        Change within noise threshold.
Found 11 outliers among 100 measurements (11.00%)
  6 (6.00%) high mild
  5 (5.00%) high severe

```
</details>

**Comment by zzlk** [comment]
also maybe TODO: merge set/multi impl and lattice_join impl too.

File: hydroflow/src/compiled/pull/half_join_state/multiset.rs [hunk 1]
```diff
@@ -5,10 +5,18 @@ use std::collections::{hash_map::Entry, VecDeque};
 
 type HashMap<K, V> = rustc_hash::FxHashMap<K, V>;
 
+use smallvec::smallvec;
+use smallvec::SmallVec;
```
Comment (MingweiSamuel) [inline hydroflow/src/compiled/pull/half_join_state/multiset.rs:9]: Nit: Combine the two uses

File: hydroflow/src/compiled/pull/half_join_state/set.rs [hunk 2]
```diff
@@ -5,10 +5,18 @@ use std::collections::{hash_map::Entry, VecDeque};
 
 type HashMap<K, V> = rustc_hash::FxHashMap<K, V>;
 
+use smallvec::smallvec;
+use smallvec::SmallVec;
```
Comment (MingweiSamuel) [inline hydroflow/src/compiled/pull/half_join_state/set.rs:9]: Nit

**Review by MingweiSamuel — APPROVED** [review]
Seems like a good win

File: hydroflow/src/compiled/pull/half_join_state/multiset.rs (hunk 1 above)
Comment (zzlk) [reply hydroflow/src/compiled/pull/half_join_state/multiset.rs:9]: I wonder why cargo fmt didn't combine them or complain about it?

**Pull request merged**